File: src/main/java/raylras/zen/langserver/ZenLanguageService.java
package raylras.zen.langserver;

import org.eclipse.lsp4j.*;
import org.eclipse.lsp4j.jsonrpc.messages.Either;
import org.eclipse.lsp4j.services.LanguageClient;
import org.eclipse.lsp4j.services.TextDocumentService;
import org.eclipse.lsp4j.services.WorkspaceService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import raylras.zen.code.CompilationUnit;
import raylras.zen.langserver.provider.*;
import raylras.zen.util.Compilations;
import raylras.zen.util.PathUtils;

import java.nio.file.Path;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class ZenLanguageService implements TextDocumentService, WorkspaceService {

    private static final Logger logger = LoggerFactory.getLogger(ZenLanguageService.class);

    private static LanguageClient client;
    private final WorkspaceManager workspaceManager;

    public ZenLanguageService() {
        this.workspaceManager = new WorkspaceManager();
    }

    /* Text Document Service */

    @Override
    public void didOpen(DidOpenTextDocumentParams params) {
        try {
            workspaceManager.createEnvIfNotExists(params.getTextDocument().getUri());
        } catch (Exception e) {
            logger.error("Failed to process 'didOpen' event: {}", params, e);
        }
    }

    @Override
    public void didChange(DidChangeTextDocumentParams params) {
        try (Document doc = workspaceManager.openAsWrite(params.getTextDocument())){
            String source = params.getContentChanges().get(0).getText();
            Compilations.reload(doc, source);
        } catch (Exception e) {
            logger.error("Failed to process 'didChange' event: {}", params, e);
        }
    }

    @Override
    public void didClose(DidCloseTextDocumentParams params) {
    }

    @Override
    public void didSave(DidSaveTextDocumentParams params) {
    }

    @Override
    public CompletableFuture<SemanticTokens> semanticTokensFull(SemanticTokensParams params) {
        try (Document doc = workspaceManager.openAsRead(params.getTextDocument())){
            return SemanticTokensProvider.semanticTokensFull(doc, params);
        } catch (Exception e) {
            logger.error("Failed to process 'semanticTokensFull' request: {}", params, e);
            return SemanticTokensProvider.empty();
        }
    }

    @Override
    public CompletableFuture<Hover> hover(HoverParams params) {
        try (Document doc = workspaceManager.openAsRead(params.getTextDocument())) {
            return HoverProvider.hover(doc, params);
        } catch (Exception e) {
            logger.error("Failed to process 'hover' request: {}", params, e);
            return HoverProvider.empty();
        }
    }

    @Override
    public CompletableFuture<Either<List<? extends Location>, List<? extends LocationLink>>> definition(DefinitionParams params) {
        try (Document doc = workspaceManager.openAsRead(params.getTextDocument())){
            return DefinitionProvider.definition(doc, params);
        } catch (Exception e) {
            logger.error("Failed to process 'definition' request: {}", params, e);
            return DefinitionProvider.empty();
        }
    }

    @Override
    public CompletableFuture<List<? extends Location>> references(ReferenceParams params) {
        try (Document doc = workspaceManager.openAsRead(params.getTextDocument())){
            return ReferencesProvider.references(doc, params);
        } catch (Exception e) {
            logger.error("Failed to process 'documentSymbol' request: {}", params, e);
            return ReferencesProvider.empty();
        }
    }

    @Override
    public CompletableFuture<List<Either<SymbolInformation, DocumentSymbol>>> documentSymbol(DocumentSymbolParams params) {
        try (Document doc = workspaceManager.openAsRead(params.getTextDocument())) {
            return DocumentSymbolProvider.documentSymbol(doc, params);
        } catch (Exception e) {
            logger.error("Failed to process 'documentSymbol' request: {}", params, e);
            return DocumentSymbolProvider.empty();
        }
    }

    @Override
    public CompletableFuture<Either<List<CompletionItem>, CompletionList>> completion(CompletionParams params) {
        try (Document doc = workspaceManager.openAsRead(params.getTextDocument())) {
            return CompletionProvider.completion(doc, params);
        } catch (Exception e) {
            logger.error("Failed to process 'completion' request: {}", params, e);
            return CompletionProvider.empty();
        }
    }

    @Override
    public CompletableFuture<CompletionItem> resolveCompletionItem(CompletionItem unresolved) {
        return CompletableFuture.completedFuture(null);
    }

    /* End Text Document Service */

    /* Workspace Service */

    @Override
    public void didChangeConfiguration(DidChangeConfigurationParams params) {
    }

    @Override
    public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
        params.getChanges().forEach(event -> {
            try {
                workspaceManager.createEnvIfNotExists(event.getUri());
                workspaceManager.getEnv(event.getUri()).ifPresent(env -> {
                    Path documentPath = PathUtils.toPath(event.getUri());
                    switch (event.getType()) {
                        case Created -> {
                            CompilationUnit unit = env.createUnit(documentPath);
                            Compilations.loadUnit(unit);
                        }
                        case Changed -> {
                            CompilationUnit unit = env.getUnit(documentPath);
                            Compilations.loadUnit(unit);
                        }
                        case Deleted -> {
                            env.removeUnit(documentPath);
                        }
                    }
                });
            } catch (Exception e) {
                logger.error("Failed to process 'didChangeWatchedFiles' event: {}", event, e);
            }
        });
    }

    @Override
    public void didChangeWorkspaceFolders(DidChangeWorkspaceFoldersParams params) {
        params.getEvent().getRemoved().forEach(workspace -> {
            workspaceManager.removeWorkspace(workspace);
            logger.info("Workspace folder removed: {}", workspace);
        });
        params.getEvent().getAdded().forEach(workspace -> {
            workspaceManager.addWorkspace(workspace);
            logger.info("Workspace folder added: {}", workspace);
        });
    }

    /* End Workspace Service */

    public static LanguageClient getClient() {
        return client;
    }

    public static void setClient(LanguageClient client) {
        ZenLanguageService.client = client;
    }

    public static void showMessage(MessageParams messageParams) {
        if (client != null) {
            client.showMessage(messageParams);
        }
    }

    public void initializeWorkspaces(List<WorkspaceFolder> workspaces) {
        if (workspaces != null) {
            workspaces.forEach(workspace -> {
                workspaceManager.addWorkspace(workspace);
                logger.info("{}", workspace);
            });
        }
    }

}

File: language-server/src/main/kotlin/raylras/intellizen/lsp/ZenLanguageService.kt
package raylras.intellizen.lsp

import org.eclipse.lsp4j.CompletionItem
import org.eclipse.lsp4j.CompletionList
import org.eclipse.lsp4j.CompletionParams
import org.eclipse.lsp4j.DefinitionParams
import org.eclipse.lsp4j.DidChangeConfigurationParams
import org.eclipse.lsp4j.DidChangeTextDocumentParams
import org.eclipse.lsp4j.DidChangeWatchedFilesParams
import org.eclipse.lsp4j.DidCloseTextDocumentParams
import org.eclipse.lsp4j.DidOpenTextDocumentParams
import org.eclipse.lsp4j.DidSaveTextDocumentParams
import org.eclipse.lsp4j.DocumentSymbol
import org.eclipse.lsp4j.DocumentSymbolParams
import org.eclipse.lsp4j.Hover
import org.eclipse.lsp4j.HoverParams
import org.eclipse.lsp4j.InlayHint
import org.eclipse.lsp4j.InlayHintParams
import org.eclipse.lsp4j.Location
import org.eclipse.lsp4j.LocationLink
import org.eclipse.lsp4j.ReferenceParams
import org.eclipse.lsp4j.SemanticTokens
import org.eclipse.lsp4j.SemanticTokensParams
import org.eclipse.lsp4j.SemanticTokensRangeParams
import org.eclipse.lsp4j.SymbolInformation
import org.eclipse.lsp4j.jsonrpc.messages.Either
import org.eclipse.lsp4j.services.TextDocumentService
import org.eclipse.lsp4j.services.WorkspaceService
import raylras.intellizen.lsp.provider.CompletionProvider
import raylras.intellizen.lsp.provider.DefinitionProvider
import raylras.intellizen.lsp.provider.DocumentSymbolProvider
import raylras.intellizen.lsp.provider.HoverProvider
import raylras.intellizen.lsp.provider.InlayHintProvider
import raylras.intellizen.lsp.provider.ReferencesProvider
import raylras.intellizen.lsp.provider.SemanticTokensProvider
import java.util.concurrent.CompletableFuture

class ZenLanguageService : TextDocumentService, WorkspaceService {
    override fun didOpen(params: DidOpenTextDocumentParams) {

    }

    override fun didChange(params: DidChangeTextDocumentParams) {

    }

    override fun didClose(params: DidCloseTextDocumentParams) {

    }

    override fun didSave(params: DidSaveTextDocumentParams) {

    }

    override fun didChangeConfiguration(params: DidChangeConfigurationParams) {

    }

    override fun didChangeWatchedFiles(params: DidChangeWatchedFilesParams?) {

    }

    override fun completion(params: CompletionParams): CompletableFuture<Either<List<CompletionItem>, CompletionList>> {
        val result = CompletionProvider.completion(params)
        return CompletableFuture.completedFuture(Either.forRight(result))
    }

    override fun resolveCompletionItem(unresolved: CompletionItem): CompletableFuture<CompletionItem> {
        val result = CompletionProvider.resolveCompletionItem(unresolved)
        return CompletableFuture.completedFuture(result)
    }

    override fun definition(params: DefinitionParams): CompletableFuture<Either<List<Location>, List<LocationLink>>> {
        val result = DefinitionProvider.definition(params)
        return CompletableFuture.completedFuture(Either.forRight(result))
    }

    override fun documentSymbol(params: DocumentSymbolParams): CompletableFuture<List<Either<SymbolInformation, DocumentSymbol>>> {
        val result = DocumentSymbolProvider.documentSymbol(params)
        return CompletableFuture.completedFuture(result.map { Either.forRight<SymbolInformation, DocumentSymbol>(it) })
    }

    override fun hover(params: HoverParams): CompletableFuture<Hover?> {
        val result = HoverProvider.hover(params)
        return CompletableFuture.completedFuture(result)
    }

    override fun inlayHint(params: InlayHintParams): CompletableFuture<List<InlayHint>> {
        val result = InlayHintProvider.inlayHint(params)
        return CompletableFuture.completedFuture(result)
    }

    override fun references(params: ReferenceParams): CompletableFuture<List<Location>> {
        val result = ReferencesProvider.references(params)
        return CompletableFuture.completedFuture(result)
    }

    override fun semanticTokensFull(params: SemanticTokensParams): CompletableFuture<SemanticTokens> {
        val result = SemanticTokensProvider.semanticTokensFull(params)
        return CompletableFuture.completedFuture(result)
    }

    override fun semanticTokensRange(params: SemanticTokensRangeParams): CompletableFuture<SemanticTokens> {
        val result = SemanticTokensProvider.semanticTokensRange(params)
        return CompletableFuture.completedFuture(result)
    }
}
